File: hackrank-apr-math/kelement.py

MOD = 10**9 + 7
table = [0] * 1000005

def inverse(a, mod):
	t, newt = 0, 1
	r, newr = mod, a;
	while newr != 0:
		q = r / newr
		t, newt = newt, t - q * newt
		r, newr = newr, r - q * newr
	if t < 0:
		t += mod
	return t

def preprocess():
	for k in xrange(1, 1000005):
		table[k] = inverse(k, MOD)

def solve(n, k):
	if k == 0 or k == n:
		return 1
	k = min(k, n-k)
	res = 1
	for i in xrange(k):
		res = (res * (n - i)) % MOD
		res = (res * table[i+1]) % MOD
	return res

if __name__ == "__main__":
	preprocess()
	t = int(raw_input())
	for _ in xrange(t):
		[n, k] = map(int, raw_input().split())
		print solve(n-1, k-1)

File: hackrank-apr-math/kelement.cpp
#include <cstdio>
#include <algorithm>

typedef long long LL;

static const int N = 2000001;
static const LL MOD = 1e9 + 7;
LL fac[N];

LL inverse(const LL& a) {
	LL r = MOD, newr = a;
	LL t = 0, newt = 1;
	LL q, temp;
	while (newr != 0) {
		q = r / newr;
		temp = r - q * newr;
		r = newr;
		newr = temp;
		temp = t - q * newt;
		t = newt;
		newt = temp;
	}
	return t < 0 ? t + MOD : t;
}

void process() {
	fac[1] = 1;
	for (LL i = 2; i < N; ++i) {
		fac[i] = (fac[i-1] * i) % MOD;
	}
}

LL solve(const LL& n, const LL& k) {
	if (k == n || k == 0)
		return 1;
	LL res = ((fac[n] * inverse(fac[n - k])) % MOD * inverse(fac[k])) % MOD;
	return res;
}

int main() {
	process();
	int t;
	scanf("%d", &t);
	LL n, k;
	while (t--) {
		scanf("%lld%lld", &n, &k);
		printf("%lld\n", solve(n-1, k-1));
	}
	return 0;
}